agent/lib/kontena/service_pods/infra_manager.rb

```diff
@@ -0,0 +1,76 @@
+module Kontena
+  module ServicePods
+    class InfraManager
+      include Kontena::Logging
+      include Kontena::Helpers::WeaveHelper
+      include Common
+
+      CONTAINER_TYPE = 'infra'.freeze
+
+      attr_reader :service_pod
+
+      # @param service_pod [ServicePod]
+      def initialize(service_pod)
+        @service_pod = service_pod
+      end
+
+      # @param service_container [Docker::Container,NilClass]
+      # @return [Container]
+      def ensure_infra(service_container = nil)
+        infra_container = get_container(service_pod.service_id, service_pod.instance_number, CONTAINER_TYPE)
+        unless infra_container
+          remove_container(service_container) if service_container
+          infra_container = create_infra
+        end
+
+        start_infra(infra_container) unless infra_container.running?
+
+        infra_container
+      end
+
+      # @return [Docker::Container]
+      def create_infra
+        network_adapter.modify_create_opts(config) if service_pod.can_expose_ports?
+        info "creating infra for service: #{service_pod.name_for_humans}"
+        Docker::Container.create(config)
+      end
+
+      # @param infra_container [Docker::Container]
+      def start_infra(infra_container)
+        infra_container.start!
+        attach_network(infra_container) if infra_container.overlay_network && infra_container.overlay_cidr
+      end
+
+      def terminate
+        infra_container = get_container(service_pod.service_id, service_pod.instance_number, CONTAINER_TYPE)
+        if infra_container
+          detach_network(infra_container) if infra_container.overlay_network && infra_container.overlay_cidr
+          remove_container(infra_container)
```

Comment:
Safer to first `remove_container` and then `detach_network`. Avoid ending up with the IPAM address released but the container still alive. The other way around is handled by the routine IPAM cleanup.